README.rst
django-supercaptcha
===================

Captchafield for Django (based on `supercaptcha`_)


Install
=======

.. code-block:: bash

  pip install django-supercaptcha

Add ``django-supercaptcha`` to INSTALLED_APPS:

.. code-block:: python

  INSTALLED_APPS = (
      'django-supercaptcha',
  )

.. important::
    `Pillow`_ must be installed


Usage
=====

.. code-block:: python

    class MySuperForm(forms.Form):
        captcha = django_supercaptcha.CaptchaField(label=u"no robots here")


All you need, to get captcha working is properly configured cache backend and a route:


.. code-block:: python

    url(r'^captcha/(?P<code>[\da-f]{32})/$', 'django_supercaptcha.draw')



Options
=======

.. list-table::
  :widths: 10 20 5

  * - CAPTCHA_SYMBOLS
    - string with all symbols, which should be used in captcha.
    - ``default``: ``123456789ABCDEFGHJKLMNPQRSTVXYZ``
  * - CAPTCHA_LENGTH
    - length if captcha code.
    - ``default``: ``4``
  * - CAPTCHA_FONTS
    - the tuple of tuples with name - fullpath pair for each font which should be used
    -
  * - CAPTCHA_FOREGROUND_COLORS
    - tuple of tuples with colors for text. use one if if you want single color, for example ((0,0,0),) for black color
    -
  * - CAPTCHA_BACKGROUND_COLOR`
    - background color,
    - ``deafult``: ``(255, 255, 255)``
  * - CAPTCHA_FILTER_CHAIN
    - PIL filters, for example `('BLUR', 'SHARPEN',)`
    - ``default``: ``[]``
  * - CAPTCHA_VERTICAL_JUMP
    - defines if letters should "jump",
    - ``default``: ``True``
  * - CAPTCHA_SIZE
    - tuple defining size of captcha image
    - ``default`: ``(120, 50)``
  * - CAPTCHA_ALT
    - "alt" for image tag,
    - ``default``: ``no robots here``
  * - CAPTCHA_TITLE
    - "title" for image tag,
    - ``default``: ``no robots here``
  * - CAPTCHA_FORMAT
    - controls which format will be used for image encoding
    - ``default``: ``('JPEG', 'image/jpeg')``
  * - CAPTCHA_CACHE_PREFIX
    - defines which prefix should supercaptcha use for dealing with cache
    - ``default``: ``captcha``
  * - CAPTCHA_DEFAULT_ERROR_MESSAGE
    - defines default error message for wrong code
    - ``default``: ``The code you entered is wrong``
  * - CAPTCHA_REFRESH
    - defines if CaptchaField should show refresh link
    - ``default``: ``True``
  * - CAPTCHA_REFRESH_LINK_TEXT`
    - defines text of refresh link
    - ``default``: ``refresh?``
  * - CAPTCHA_HTML_TEMPLATE`
    - defines template of ``CaptchaField``, see example in settings ``settings.py``
    -
  * - CAPTCHA_HTML_TEMPLATE_WITH_REFRESH`
    - defines template of ``CaptchaField`` with refresh link, see example in ``settings.py``
    -

.. _django: http://djangoproject.com/
.. _supercaptcha: https://github.com/barbuza/supercaptcha/
.. _pillow: http://python-pillow.org/

CHANGELOG.rst
Changelog
=========

1.0.0 (31/08/2016)
------------------
* Init

django_supercaptcha/__init__.py
# -*- coding: utf-8 -*-

import os
from random import choice, random

from PIL import Image, ImageDraw, ImageFont, ImageFilter
from django import forms
from django.core.cache import cache
from django.core.urlresolvers import reverse
from django.forms.utils import flatatt
from django.http import HttpResponse
from django.utils.safestring import mark_safe
from django.utils.translation import ugettext_lazy
from django.views.decorators.cache import never_cache

from django_supercaptcha import settings


__version__ = '1.0.0'

try:
    from threading import local
except ImportError:
    from django.utils._threading_local import local

_thread_locals = local()


WIDTH = settings.SIZE[0]
HEIGHT = settings.SIZE[1]
SYMBOLS = ugettext_lazy(settings.SYMBOLS)
LENGTH = settings.LENGTH
BG_COLOR = settings.BACKGROUND_COLOR
FG_COLORS = settings.FOREGROUND_COLORS
ENC_TYPE, MIME_TYPE = settings.FORMAT
JUMP = settings.VERTICAL_JUMP
PREFIX = settings.CACHE_PREFIX
CODE_ATTR_NAME = '_captcha_code'
ERROR_MESSAGE = settings.DEFAULT_ERROR_MESSAGE
HTML_TEMPLATE = settings.HTML_TEMPLATE
HTML_TEMPLATE_WITH_REFRESH = settings.HTML_TEMPLATE_WITH_REFRESH
REFRESH = settings.REFRESH
REFRESH_LINK_TEXT = ugettext_lazy(settings.REFRESH_LINK_TEXT)


def get_current_code():
    if not hasattr(_thread_locals, CODE_ATTR_NAME):
        code = os.urandom(16).encode('hex')
        setattr(_thread_locals, CODE_ATTR_NAME, code)
    return getattr(_thread_locals, CODE_ATTR_NAME)


def set_current_code(value):
    setattr(_thread_locals, CODE_ATTR_NAME, value)


def empty_current_code():
    if hasattr(_thread_locals, CODE_ATTR_NAME):
        delattr(_thread_locals, CODE_ATTR_NAME)


def generate_text():
    return ''.join([choice(SYMBOLS) for _ in range(LENGTH)])


@never_cache
def draw(request, code):

    font_name, fontfile = choice(settings.AVAIL_FONTS)
    cache_name = '%s-%s-size' % (PREFIX, font_name)
    text = generate_text()
    cache.set('%s-%s' % (PREFIX, code), text, 600)

    def fits(font_size):
        font = ImageFont.truetype(fontfile, font_size)
        size = font.getsize(text)
        return size[0] < WIDTH and size[1] < HEIGHT

    font_size = cache.get(cache_name, 10)
    if fits(font_size):
        while True:
            font_size += 1
            if not fits(font_size):
                font_size -= 1
                break
    else:
        while True:
            font_size -= 1
            if fits(font_size):
                break
    cache.set(cache_name, font_size, 600)

    font = ImageFont.truetype(fontfile, font_size)
    text_size = font.getsize(text)
    icolor = 'RGB'
    if len(BG_COLOR) == 4:
        icolor = 'RGBA'
    im = Image.new(icolor, (WIDTH, HEIGHT), BG_COLOR)
    d = ImageDraw.Draw(im)
    if JUMP:
        position = [(WIDTH - text_size[0]) / 2, 0]
        shift_max = HEIGHT - text_size[1]
        shift_min = shift_max / 4
        shift_max = shift_max * 3 / 4
        for char in text:
            l_size = font.getsize(char)
            try:
                position[1] = choice(range(shift_min, shift_max + 1))
            except IndexError:
                position[1] = shift_min
            d.text(position, char, font=font, fill=choice(FG_COLORS))
            position[0] += l_size[0]
    else:
        position = [(WIDTH - text_size[0]) / 2,
                    (HEIGHT - text_size[1]) / 2]
        d.text(position, text, font=font, fill=choice(FG_COLORS))

    response = HttpResponse(content_type=MIME_TYPE)

    response['cache-control'] = ('no-store, no-cache, must-revalidate,'
                                 'proxy-revalidate')

    for f in settings.FILTER_CHAIN:
        im = im.filter(getattr(ImageFilter, f))

    im.save(response, ENC_TYPE)
    return response


class CaptchaImageWidget(forms.Widget):

    if REFRESH:
        template = HTML_TEMPLATE_WITH_REFRESH
    else:
        template = HTML_TEMPLATE

    def render(self, name, value, attrs=None):
        code = get_current_code()
        empty_current_code()
        input_attrs = self.build_attrs(attrs, type='text', name=name)
        template_kwargs = {
            'src': reverse(draw, kwargs={'code': code}),
            'input_attrs': flatatt(input_attrs),
            'alt': settings.ALT,
            'title': settings.TITLE,
            'width': WIDTH,
            'length': LENGTH,
            'height': HEIGHT,
            'rnd': random(),
            'refresh_text': REFRESH_LINK_TEXT
        }
        return mark_safe(self.template % template_kwargs)


class HiddenCodeWidget(forms.HiddenInput):

    def render(self, name, value=None, attrs=None):
        if value is None:
            empty_current_code()
        if not value:
            value = get_current_code()
        else:
            set_current_code(value)
        return super(HiddenCodeWidget, self).render(name, value, attrs=attrs)


class CaptchaWidget(forms.MultiWidget):

    def __init__(self, attrs={}, code=None):
        widgets = (
            HiddenCodeWidget(attrs=attrs),
            CaptchaImageWidget(attrs=attrs)
        )
        super(CaptchaWidget, self).__init__(widgets, attrs)

    def decompress(self, value):
        if value:
            return value.split()
        return [None, None]

    @classmethod
    def id_for_label(cls, id_):
        if id_:
            id_ += '_1'
        return id_


class CaptchaField(forms.MultiValueField):

    widget = CaptchaWidget

    default_error_messages = {
        'wrong': ugettext_lazy(ERROR_MESSAGE),
        'required': ugettext_lazy(u'This field is required.'),
        'internal': ugettext_lazy(u'Internal error.'),
        }

    def __init__(self, *args, **kwargs):
        fields = (
            forms.CharField(max_length=32, min_length=32),
            forms.CharField(max_length=settings.LENGTH,
                            min_length=settings.LENGTH),
        )
        super(CaptchaField, self).__init__(fields, *args, **kwargs)

    def compress(self, data_list):
        return ' '.join(data_list)

    def clean(self, value):
        if len(value) != 2:
            raise forms.ValidationError(self.error_messages['wrong'])

        code, text = value
        if not text:
            raise forms.ValidationError(self.error_messages['required'])

        cached_text = cache.get('%s-%s' % (PREFIX, code))
        cache.set('%s-%s' % (PREFIX, code), generate_text(), 600)
        if not cached_text:
            raise forms.ValidationError(self.error_messages['internal'])

        if text.lower() != cached_text.lower():
            raise forms.ValidationError(self.error_messages['wrong'])


django_supercaptcha/settings.py
# -*- coding: utf-8 -*-

from os.path import join, dirname, abspath

from django.conf import settings


SELF_DIR = dirname(abspath(__file__))
FONTS_DIR = join(SELF_DIR, 'fonts')

SYMBOLS = getattr(settings, 'CAPTCHA_SYMBOLS',
                  '123456789ABCDEFGHJKLMNPQRSTVXYZ')
LENGTH = getattr(settings, 'CAPTCHA_LENGTH', 4)

AVAIL_FONTS = getattr(settings, 'CAPTCHA_FONTS', [
        ('boneca', join(FONTS_DIR, 'boneca.ttf')),
        ('acidic', join(FONTS_DIR, 'acidic.ttf')),
])

FOREGROUND_COLORS = getattr(settings, 'CAPTCHA_FOREGROUND_COLORS', (
    (0, 0, 0),
    (0x77, 0, 0),
    (0, 0x77, 0),
    (0, 0, 0x77),
))

COLORIZE_SYMBOLS = getattr(settings, 'CAPTCHA_COLORIZE_SYMBOLS', True)

BACKGROUND_COLOR = getattr(settings, 'CAPTCHA_BACKGROUND_COLOR',
                           (255, 255, 255))

FILTER_CHAIN = getattr(settings, 'CAPTCHA_FILTER_CHAIN', [])

VERTICAL_JUMP = getattr(settings, 'CAPTCHA_VERTICAL_JUMP', True)

SIZE = getattr(settings, 'CAPTCHA_SIZE', (120, 50))
ALT = getattr(settings, 'CAPTCHA_ALT', 'No robots here')
TITLE = getattr(settings, 'CAPTCHA_TITLE', 'No robots here')
FORMAT = getattr(settings, 'CAPTCHA_FORMAT', ('JPEG', 'image/jpeg'))

CACHE_PREFIX = getattr(settings, 'CAPTCHA_CACHE_PREFIX', 'captcha')

DEFAULT_ERROR_MESSAGE = getattr(settings, 'CAPTCHA_DEFAULT_ERROR_MESSAGE',
                                u'The code you entered is wrong')

REFRESH_LINK_TEXT = getattr(settings, 'CAPTCHA_REFRESH_LINK_TEXT',
                            u'refresh?')

REFRESH = getattr(settings, 'CAPTCHA_REFRESH', True)

HTML_TEMPLATE = getattr(
    settings,
    'CAPTCHA_HTML_TEMPLATE',
    """
    <div class="captcha">
        <div class="captcha-image">
            <span class="captcha-image__image">
                <img id="captcha_field_image"
                     width="%(width)s"
                     height="%(height)s"
                     src="%(src)s?%(rnd)s"
                     title="%(title)s"
                     alt="%(alt)s" />
            </span>
        </div>
        <div class="captcha-input">
            <input%(input_attrs)s maxlength="%(length)s"
                   class="captcha-input__input" />
        </div>
    </div>
    """
)

HTML_TEMPLATE_WITH_REFRESH = getattr(
    settings,
    'CAPTCHA_HTML_TEMPLATE_WITH_REFRESH',
    """
    <div class="captcha">
        <div class="captcha-image">
            <span class="captcha-image__image">
                <img id="captcha_field_image"
                     width="%(width)s"
                     height="%(height)s"
                     src="%(src)s?%(rnd)s"
                     title="%(title)s"
                     alt="%(alt)s" />
            </span>
            <a href="#refresh" class="captcha__link" onclick="var
               img=document.getElementById('captcha_field_image');
               img.src=img.src.substring(0,img.src.indexOf('?')+1)+Math.random();
               return false;">%(refresh_text)s</a>
        </div>
        <div class="captcha-input">
            <input%(input_attrs)s maxlength="%(length)s"
                   class="captcha-input__input" />
        </div>
    </div>
    """
)
